src/solvers/twostep_time_solver.C

```diff
@@ -339,4 +340,60 @@ void TwostepTimeSolver::integrate_adjoint_sensitivity(const QoISet & qois, const
      sensitivities[i][j] = sensitivities_first_half[i][j] + sensitivities_second_half[i][j];
 }
 
+void TwostepTimeSolver::integrate_adjoint_refinement_error_estimate(AdjointRefinementEstimator & adjoint_refinement_error_estimator, ErrorVector & QoI_elementwise_error, std::vector<Real *> QoI_time_instant)
+{
+  // We are using the midpoint rule or right sided Riemman rule to integrate each timestep
+  // (f(t_j) + f(t_j+1/2))/2 (t_j+1/2 - t_j) + (f(t_j+1/2) + f(t_j+1))/2 (t_j+1 - t_j+1/2)
+
+  // Create first and second half error estimate vectors of the right size
+  std::vector<Number> qoi_error_estimates_first_half(_system.qoi.size());
+  std::vector<Number> qoi_error_estimates_second_half(_system.qoi.size());
+
+  // First half timestep
+  ErrorVector QoI_elementwise_error_first_half;
+
+  core_time_solver->integrate_adjoint_refinement_error_estimate(adjoint_refinement_error_estimator, QoI_elementwise_error_first_half, QoI_time_instant);
+
+  // Also get the first 'half step' spatially integrated errors for all the QoIs in the QoI set
+  for (auto j : make_range(_system.n_qois()))
+  {
+    // Skip this QoI if not in the QoI Set
+    if (adjoint_refinement_error_estimator.qoi_set().has_index(j))
+    {
+      qoi_error_estimates_first_half[j] = (_system.qoi_error_estimates)[j];
+    }
+  }
+
+  // Second half timestep
+  ErrorVector QoI_elementwise_error_second_half;
+
+  core_time_solver->integrate_adjoint_refinement_error_estimate(adjoint_refinement_error_estimator, QoI_elementwise_error_second_half, QoI_time_instant);
+
+  // Also get the second 'half step' spatially integrated errors for all the QoIs in the QoI set
+  for (auto j : make_range(_system.n_qois()))
+  {
+    // Skip this QoI if not in the QoI Set
+    if (adjoint_refinement_error_estimator.qoi_set().has_index(j))
+    {
+      qoi_error_estimates_second_half[j] = (_system.qoi_error_estimates)[j];
+    }
+  }
+
+  // Error contribution from this timestep
+  for(unsigned int i = 0; i < QoI_elementwise_error.size(); i++)
+    QoI_elementwise_error[i] = QoI_elementwise_error_first_half[i] + QoI_elementwise_error_second_half[i];
+
+  // Add the spatially integrated errors for all the QoIs in the QoI set
+  Real spatially_integrated_qoi_error;
+
+  for (auto j : make_range(_system.n_qois()))
+  {
+    // Skip this QoI if not in the QoI Set
+    if (adjoint_refinement_error_estimator.qoi_set().has_index(j))
+    {
+      (_system.qoi_error_estimates)[j] = qoi_error_estimates_first_half[j] + qoi_error_estimates_second_half[j];
```

Comment:
There's got to be a better nomenclature we can use here.  Correct me if I'm wrong, but:

```qoi_error_estimates``` isn't an estimate of the QoI error, but only the component of that estimate which comes from a single time step?

```integrate_adjoint_refinement_error_estimate``` integrates it over the current time step (or two sub time steps), but doesn't add it to the time integration in progress?

Those are defensible choices but the names aren't self-documenting, they're almost anti-self-documenting.  And regardless of name we're going to need much, much more documentation in the headers.

Reply:
I still haven't come up with much better names, so we can leave it.  And the qoi_error_estimates documentation in system.h is clear enough.  Can you edit the integrate_adjoint_refinement_error_estimate doc to say something clearer than "A method to compute the adjoint refinement error estimate at the current timestep", though?  We're not actually estimating the error here, we're estimating the contribution of the error from the current timestep, right?  That needs to be absolutely clear in the doxygen.